File: algorithms/144. Binary Tree Preorder Traversal/solution3.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> preorderTraversal(TreeNode* root) {
        stack<TreeNode *> s;
        vector<int> values;
        while ( root || s.size() > 0 ) {
            while ( root ) {
                values.push_back(root->val);
                s.push(root);
                root = root->left;
            }
            root = s.top();
            s.pop();
            root = root->right;
        }
        return values;
    }
};

File: algorithms/106. Construct Binary Tree from Inorder and Postorder Traversal/solution3.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        if inorder:
            p = inorder.index(postorder.pop())
            root = TreeNode(inorder[p])
            root.right = self.buildTree(inorder[p+1:], postorder)
            root.left = self.buildTree(inorder[:p], postorder)
            return root

# 202 / 202 test cases passed.
# Status: Accepted
# Runtime: 212 ms
# beats 68.81 %
